File: 47-Permutations II.cpp
//47-Permutations II.cpp

#include<iostream>
#include<vector>
using namespace std;

//20170325
//Time:O(N^a)
//Space:O(1)
//backtracking
class Solution{
public:
    vector<vector<int>> permuteUnique(vector<int>& nums)
    {
        vector<vector<int>> r={};
        backTracking(0,nums.size(),r,nums);
        return r;
    }
    void backTracking(int t,int n,vector<vector<int>>& r,vector<int>& nums)
    {
        if(t==n){
            // vector<int> t;
            // for(int i=0;i<n;i++){
            //     t.push_back(nums[i]);
            // }
            // if(!isExit(r,t))
            //     r.push_back(t);
            if(!isExit(r,nums))
            	r.push_back(nums);
        }
        else{
            for(int i=t;i<n;i++){
                swap(nums[i],nums[t]);
                backTracking(t+1,nums.size(),r,nums);
                swap(nums[i],nums[t]);
            }
        }

    }
    bool isExit(vector<vector<int>>& r,vector<int>& t)
    {
        for(int k=0;k<r.size();k++){
            if(r[k]==t)
                return true;
        }
        return false;

    }
};

int main()
{
    Solution s;
    vector<int> v={};
    vector<vector<int>> r=s.permuteUnique(v);
    for(int i=0;i<r.size();i++){
        for(int k=0;k<r[i].size();k++)
            cout<<r[i][k];
        cout<<endl;
    }

    return 0;
}

File: 47. Permutations II.py
# -*- coding: utf-8 -*-
# @Time    : 2019/2/12
# @Author  : qirui
# @FileName: 47. Permutations II.py

# reference: https://leetcode.com/problems/combination-sum/discuss/16502/A-general-approach-to-backtracking-questions-in-Java-(Subsets-Permutations-Combination-Sum-Palindrome-Partitioning)
# Backtrack/DFS (general)
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        def backtrack(tmp):
            if len(tmp) == N:
                res.append(tmp)
                return
            for i in range(N):
                if used[i] or (i > 0 and not used[i - 1] and nums[i] == nums[i - 1]):   continue
                used[i] = True
                backtrack(tmp + [nums[i]])
                used[i] = False

        N = len(nums)
        res = []
        used = [False for _ in range(N)]
        nums.sort()
        backtrack([])
        return res


if __name__ == '__main__':
    S = Solution()
    print(S.permuteUnique([1, 1, 2]))
